### PR TITLE
test: add integration tests

### PR DESCRIPTION
**Pull request by brettstack**
*Description of changes:*
Adds integration tests. Brings our coverage to 98%.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
 node_modules
 npm-debug.log
 .DS_Store
-.env
\ No newline at end of file
+.env
+coverage
```
Comment (brettstack) [inline .gitignore]: Don't want to include coverage data

File: .travis.yml [hunk 2]
```diff
@@ -4,7 +4,12 @@ node_js:
   - 4
   - 6
   - 8
-  - node # Latest
+  # - node # Latest; NOTE: latest does not work due to Buffer deprecation
+
+before_install:
+  - echo 'Installing example dependencies...'
+  - cd example && npm install && cd ..
```
Comment (brettstack) [inline .travis.yml]: The tests utilize the existing example which covers a lot of bases. Need to install the example's dependencies for it to work.

File: __tests__/integration.js [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
...dx5+7EnuQcJ+H48R/g09/6rN1OjQRlZ565vwPl5T9FWcg5Uxs4TVAXL76F0ZGN+NMnz9jx9EhQTuC5rV9FZeU/wOt1k6BswtjYGAYv/Rfc8fiP7HyaJCgn8e8P/Rrl5W3weBgSlEVJJnlcvPg0PvIPf+GEW5YE5TR+fP9SBEM/x6LyJhKUheA4CRcvvoKJ+L12TedIUMRlnv3yZxGJfst29Sk7CurixbcwNrYZH3viV067TUlQTucnDzyNReWfsU19yk6CiseTuHTxb7Dp6zudenuSoAg57fP5/tkW9Sk7CIrjJLzzznedUmciQRGF8cz9HQiF/snS9SmrC+rcuX1OqzORoIjieG7rVxEtfwTBgJ8EpRNDQxcwPPxxJ9aZSFCEMn7ywNOorPicpdI+qwkqmeQx8M5DTq4zkaAI5eTqU1VVG0hQKsJxEs6f/yk+9Pcfo5uMBEWUyjP3dyAS/hEikRoSVImcH+hDLHYn1ZlIUITamH3YjJkF5ZDhKSQownh2P7gXFYs+Yrr6lBkFlUzyOH/+cdz+2F/TjUOCIvTix/cvRSDw76iouI4ENQe54Smdj3bQzUKCIozCTMNmzCIom0+DQoIirMe//fcnULHoy4bWp4wW1NjYGIaHHqFuAyQowqwYOa2LUYJy2DQoJCjC2hg1rYvegnLoNCgkKMIePLf1qwhH/odu9Sk9BUXDU0hQhE3Qa1oXPQRF06CQoAibpn1aT+uipaCozkSCIhwkqkjkZtUjKi0ERWIiQREOTv1CoTtVq1GpKaihoQuIje2gVI4ERTidZ+7vQDDwv1FWdm1JUVWpgorHkxgdfQHJ5BNU/CZBEcRsnv3yZ+FhP4Vg8H1FR1bFCorjJMRiFzEx8TOk07tISiQoglAmLI+7EV7f1fD7wvMW2RcSVDLJI52eQDLRj3TmMLjMizSzAAmKILRLC92um2Z8z+0W4HYLM77HC4coMiJBEQRBqIqLmoAgCBIUQRAECYogCBIUQRAECYogCBIUQRAECYogCIIERRAECYogCIIERRAECYogCIIERRAEQYIiCIIERRAEQYIiCIIERRAEQYIiCIIgQREEQYIiCIIgQREEQYIiCIIgQREEQZCgCIIgQREEQZCgCIIgQREEQZCgCIIgSFAEQZCgCIIgSFAEQZCgCIIgtOf/A7sWdfziDxE6AAAAAElFTkSuQmCC",
```
Comment (brettstack) [inline __tests__/integration.js]: This is the reason this file is originally hidden due to being too large. It's the base64 encoded version of https://github.com/awslabs/aws-serverless-express/blob/master/example/sam-logo.png. A few lines down we're creating a new instance of a server which has a directive to base64 encode the response when it encounters image/* content types.

File: __tests__/integration.js [hunk 4]
```diff
@@ -0,0 +1,391 @@
+const awsServerlessExpress = require('../index')
+const apiGatewayEvent = require('../example/api-gateway-event.json')
+const app = require('../example/app')
+
+const server = awsServerlessExpress.createServer(app)
+const lambdaFunction = {
```
Comment (brettstack) [inline __tests__/integration.js]: This is our mock Lambda function handler. See https://github.com/awslabs/aws-serverless-express/blob/master/example/lambda.js and you'll see this is the core of our Lambda function. This server/"lambda function" will be used in the majority of our tests. For tests with special cases (e.g. base64 encoding images; server callback; socket path conflict error/edge case) a new server is created with special configuration.

File: __tests__/integration.js [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+  handler: (event, context) => awsServerlessExpress.proxy(server, event, context)
+}
+
+function clone(json) {
```
Comment (brettstack) [inline __tests__/integration.js]: Because JS doesn't have a native way for cloning objects 👎

File: __tests__/integration.js [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+  return JSON.parse(JSON.stringify(json))
+}
+
+function makeEvent(eventOverrides) {
```
Comment (brettstack) [inline __tests__/integration.js]: Helper function for making `event`s to reduce boilerplate. It uses the example event included in the `example` directory as a base and modifies it according to the needs of each test.

File: __tests__/integration.js [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+  const baseEvent = clone(apiGatewayEvent)
+  const headers = Object.assign({}, baseEvent.headers, eventOverrides.headers)
+  delete eventOverrides.headers
+  const root = Object.assign({}, baseEvent, eventOverrides)
+  root.headers = headers
```
Comment (brettstack) [inline __tests__/integration.js]: This headers stuff could probably be cleaned up, but since`Object.assign` does not do a deep merge, we need to first merge the `eventOverride.headers` with the base/example event headers then set those in the resulting merged event.

File: __tests__/integration.js [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+  return root
+}
+
+function expectedRootResponse() {
```
Comment (brettstack) [inline __tests__/integration.js]: Helper function for root response since we have multiple tests which hit root route. The same could probably be applied for other routes.

File: __tests__/integration.js [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+  return makeResponse({
+    "headers": {
+      "content-length": "3747",
+      "content-type": "text/html; charset=utf-8",
+      "etag": "W/\"ea3-WawLnWdlaCO/ODv9DBVcX0ZTchw\""
+    }
+  })
+}
+
+function makeResponse(response) {
```
Comment (brettstack) [inline __tests__/integration.js]: Helper function for responses. Same merging strategy with headers.

File: __tests__/integration.js [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+  const headers = Object.assign({}, baseHeaders, response.headers)
+  const finalResponse = Object.assign({}, baseResponse, response)
+  finalResponse.headers = headers
+  return finalResponse
+}
+
+describe('integration tests', () => {
+  test('GET HTML (initial request)', (done) => {
+    const succeed = response => {
+      delete response.headers.date
```
Comment (brettstack) [inline __tests__/integration.js]: date is always different. This is Express logic, not aws-serverless-express, so not important to specifically test (e.g. assert it matches a pattern).

File: __tests__/integration.js [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+      expect(response.body.startsWith('<!DOCTYPE html>')).toBe(true)
```
Comment (brettstack) [inline __tests__/integration.js]: The body here is long. Assert it starts with this is and assert content-length is enough.

File: __tests__/integration.js [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+  }
+  const headers = Object.assign({}, baseHeaders, response.headers)
+  const finalResponse = Object.assign({}, baseResponse, response)
+  finalResponse.headers = headers
+  return finalResponse
+}
+
+describe('integration tests', () => {
+  test('GET HTML (initial request)', (done) => {
+    const succeed = response => {
```
Comment (brettstack) [inline __tests__/integration.js]: The original resolution for Node.js Lambda functions is `context.succeed(response)` which this library calls. You can see on line 62/63 (and all other tests) that we specify our `context` object as the second parameter of `makeEvent`. We make all our assertions against the `response` of a `context.succeed` call

File: __tests__/integration.js [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+      delete expectedResponse.body
+      expect(response).toEqual(expectedResponse)
+      done()
+    }
+    lambdaFunction.handler(makeEvent({ path: '/', httpMethod: 'GET' }), {
+      succeed
+    })
+  })
+
+  test('GET HTML (subsequent request)', (done) => {
```
Comment (brettstack) [inline __tests__/integration.js]: aws-serverless-express has different code paths for first request (cold start) and subsequent requests. Could clean this up into a function which both of these root tests could call.

File: __tests__/integration.js [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+        statusCode: 502
+      })
+      done()
+    }
+    lambdaFunction.handler(makeEvent({
+      path: '/',
+      httpMethod: 'GET',
+      body: "{\"name\": \"Sam502\"}",
+      headers: {
+        'Content-Length': '-1'
```
Comment (brettstack) [inline __tests__/integration.js]: This causes 502

File: __tests__/integration.js [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+      expect(serverListenCallback).toHaveBeenCalled()
+      serverWithCallback.close()
+      done()
+    }
+    awsServerlessExpress.proxy(serverWithCallback, makeEvent({}), {
+      succeed
+    })
+  })
+
+  test('server.onError EADDRINUSE', (done) => {
```
Comment (brettstack) [inline __tests__/integration.js]: This error/edge case is now *incredibly* unlikely since we use a random string for the sock path.

File: __tests__/integration.js [hunk 16]
```diff
@@ -0,0 +1,391 @@
...
+      expect(response.statusCode).toBe(200)
+      done()
+      serverWithSameSocketPath.close()
+    }
+    awsServerlessExpress.proxy(serverWithSameSocketPath, makeEvent({}), {
+      succeed
+    })
+  })
+
+  test.skip('set-cookie')
```
Comment (brettstack) [inline __tests__/integration.js]: Another edge case test we can leave for another day. API Gateway does not have support for multiple headers with same name. We work around this by modifying the casing of the set-cookie header when we detect there are multiple.

File: example/lambda.js [hunk 17]
```diff
@@ -1,5 +1,5 @@
 'use strict'
-const awsServerlessExpress = require('aws-serverless-express')
+const awsServerlessExpress = require(process.env.NODE_ENV === 'test' ? '../index' : 'aws-serverless-express')
```
Comment (brettstack) [inline example/lambda.js]: Use the library version instead of the installed version from NPM during tests. This is required for:

1. Writing tests for new features
1. Coverage report

File: index.js [hunk 18]
```diff
@@ -62,7 +62,10 @@ function forwardResponseToApiGateway(server, response, context) {
             const headers = response.headers
 
             // chunked transfer not currently supported by API Gateway
-            if (headers['transfer-encoding'] === 'chunked') delete headers['transfer-encoding']
+            /* istanbul ignore else */
```
Comment (brettstack) [inline index.js]: These are just directives for the coverage reporter

File: __tests__/integration.js (hunk 3 above)
Comment (honglu) [reply __tests__/integration.js]: Question. If it is the based64 encoded string for a known file, why not just generate it instead of hardcoding it here?

Comment (brettstack) [reply __tests__/integration.js]: Good callout! I'll do that.

Comment (ombozek) [reply __tests__/integration.js]: Or store it in a txt file
